geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryMonitor.java

```diff
@@ -285,39 +258,40 @@ public static int getQueryMonitorThreadCount() {
     // package-private to avoid synthetic accessor
     final AtomicBoolean queryCancelled;
 
-    QueryThreadTask(Thread queryThread, DefaultQuery query, AtomicBoolean queryCancelled) {
-      this.StartTime = System.currentTimeMillis();
+    QueryExpiration(final Thread queryThread, final DefaultQuery query,
+        final AtomicBoolean queryCancelled,
+        final long maxQueryExecutionTime) {
+      startTime = System.currentTimeMillis();
+      expiredTime = startTime + maxQueryExecutionTime;
       this.queryThread = queryThread;
       this.query = query;
       this.queryCancelled = queryCancelled;
     }
 
-    @Override
-    public int hashCode() {
-      assert this.queryThread != null;
-      return this.queryThread.hashCode();
-    }
-
-    /**
-     * The query task in the queue is identified by the thread. To remove the task in the queue
-     * using the thread reference.
-     */
-    @Override
-    public boolean equals(Object other) {
-      if (!(other instanceof QueryThreadTask)) {
-        return false;
-      }
-      QueryThreadTask o = (QueryThreadTask) other;
-      return this.queryThread.equals(o.queryThread);
+    public long getMaxQueryExecutionTime() {
+      return expiredTime - startTime;
     }
 
     @Override
     public String toString() {
-      return new StringBuilder().append("QueryThreadTask[StartTime:").append(this.StartTime)
-          .append(", queryThread:").append(this.queryThread).append(", threadId:")
+      return new StringBuilder().append("QueryExpiration[startTime:").append(this.startTime)
+          .append(", expiredTime:").append(expiredTime).append(", queryThread:")
+          .append(this.queryThread).append(", threadId:")
           .append(this.queryThread.getId()).append(", query:").append(this.query.getQueryString())
           .append(", queryCancelled:").append(this.queryCancelled).append(']')
           .toString();
     }
+
+    @Override
+    public long getDelay(TimeUnit unit) {
+      final long diff = expiredTime - System.currentTimeMillis();
+      return unit.convert(diff, TimeUnit.MILLISECONDS);
```

Comment:
Will this the handle scenario where we have completed the query, but the timeout is a very large timeout.  Would this leave those query tasks in the queue until they completely expire, which could be a lot later?

Reply:
This is a valid point.  As part of this commit, we deleted the code that removes the entries from the queue after the query completes, because we wanted the QueryMonitor to be the sole entity which empties the queue for simplicity.  I think it is a valid concern that the queue may grow unbounded in the case where there are many short lived queries and a long expiration time.  We can solve this by adding the remove logic back in when the query completes.  We will also have to add equals() and hashcode() for removal to work properly.

Reply:
Also, the delayed method will change for each value as it gets compared since you will be using System.currentTimeMillis().  There might be some weird issues with that?

Side note:
The old way was that each query was inserted and the oldest was always at the head.  We will still be adding queries in order.

Reply:
Latest commit restores the immediate removal of the query from the queue as soon as the query completes. This addresses @jhuynh1's concern about queue growth when `MAX_QUERY_EXECUTION_TIME` is long.

We also ran some performance numbers that indicate that the new `DelayQueue` performs at least as well as the old `ConcurrentLinkedQueue`-based implementation did. We looked at time to add queries, then remove them (in the same order, and in the reverse order) and also we looked at time for the `QueryMonitor.run()` loop to process expired queries. In every case, for 1e4 to 1e5 queries in the queue, the new implementation was faster.